dqvm/src/permutations.lisp

```diff
@@ -0,0 +1,62 @@
+;;; permutations.lisp
+;;;
+;;; Author: Juan M. Bello-Rivas
+
+(in-package :dqvm2)
+
+(declaim (optimize (speed 3) (debug 0) (safety 0)))
+
+(deftype permutation () 'list)
+
+(qvm::defun-inlinable permute (permutation address)
+  (loop :for (i j) :in permutation
```

Comment:
is this function right? im suspect

Reply:
I think so. If an address a is written as 

  a = a_0 2^0 + \dotsb + a_j 2^j + \dotsb + a_k 2^k + \dotsb + a_{n-1} 2^{n-1},

then applying a qubit permutation that transposes qubits j and k should result in a permuted address 
  
  a^\prime = a_0 2^0 + \dotsb + a_k 2^j + \dotsb + a_j 2^k + \dotsb + a_{n-1} 2^{n-1}.

Also, compare

```
(qvm::map-reordered-amplitudes 0 (lambda (x y)
                                         (format t "~3d ~3d~%" x y))
                                     (nat-tuple 3 2 0 1))
  0   0
  1   2
  2   1
  3   3
  4   4
  5   6
  6   5
  7   7
  8   8
  9  10
 10   9
 11  11
 12  12
 13  14
 14  13
 15  15
nil
```
with
```
(dqvm2::print-qubit-permutation '((0 1)) 4)
   0 |0000>    0 |0000>
   1 |0001>    2 |0010>
   2 |0010>    1 |0001>
   3 |0011>    3 |0011>
   4 |0100>    4 |0100>
   5 |0101>    6 |0110>
   6 |0110>    5 |0101>
   7 |0111>    7 |0111>
   8 |1000>    8 |1000>
   9 |1001>   10 |1010>
  10 |1010>    9 |1001>
  11 |1011>   11 |1011>
  12 |1100>   12 |1100>
  13 |1101>   14 |1110>
  14 |1110>   13 |1101>
  15 |1111>   15 |1111>
```

Reply:
Ah, is your permutation a list of (i j) transposition pairs?

Reply:
That's right.

Reply:
oh then major false alarm